pkg/cmd/roachtest/tests/online_restore.go

```diff
@@ -415,3 +396,157 @@ func waitForDownloadJob(
 		}
 	}
 }
+
+// initCorrectnessRestoreSpecs initializes the restoreSpecs for correctness testing based on the
+// base restore spec by setting the workload seed, fake time, and trace file name
+func initCorrectnessRestoreSpecs(
+	t test.Test, baseSp restoreSpecs, seed uint64, fakeTime uint32, traceSuffix string,
+) (restoreSpecs, tpccRestore) {
+	t.Helper()
+	tpccWorkload, ok := baseSp.backup.workload.(tpccRestore)
+	if !ok {
+		require.Fail(t, "only tpcc workloads are supported for correctness testing")
+	}
+	baseSp.timeout = time.Duration(baseSp.timeout.Seconds()*0.45) * time.Second
+	tpccWorkload.opts.queryTraceFile = strings.ReplaceAll(baseSp.testName, "/", "-") + traceSuffix
+	if tpccWorkload.opts.seed == 0 {
+		tpccWorkload.opts.seed = seed
+	}
+	if tpccWorkload.opts.fakeTime == 0 {
+		tpccWorkload.opts.fakeTime = fakeTime
+	}
+	baseSp.backup.workload = tpccWorkload
+	return baseSp, tpccWorkload
+}
+
+type restoreStats struct {
+	restoreDriver             restoreDriver
+	collector                 clusterstats.StatCollector
+	restoreStartTime          time.Time
+	restoreEndTime            time.Time
+	downloadEndTimeLowerBound time.Time
+	workloadStartTime         time.Time
+	workloadEndTime           time.Time
+}
+
+func runRestore(
+	ctx context.Context,
+	t test.Test,
+	c cluster.Cluster,
+	sp restoreSpecs,
+	runOnline, runWorkload, useWorkarounds bool,
+) restoreStats {
+	testStartTime := timeutil.Now()
+
+	rd := makeRestoreDriver(t, c, sp)
```

Comment:
could we pull out making rd and preparing the cluster? logically, it's not part of running restore. Further, that means we don't need to bind rd to `restoreStats` which currently seems a bit odd.

Reply:
i realize this means there will be repeat code between the perf and correctness tests, but that seems fine here.

Reply:
SGTM